cairo/ethereum/cancun/keth/teardown.cairo

```diff
@@ -0,0 +1,361 @@
+    TupleLog__hash__,
+    UnionBytesLegacyTransactionEnum,
+    OptionalUnionBytesLegacyTransaction,
+    TupleUnionBytesLegacyTransaction,
+    TupleLog,
+    TupleLogStruct,
+    OptionalUnionBytesReceipt,
+    UnionBytesReceiptEnum,
+    Log,
+    LogStruct,
+)
+from ethereum.utils.numeric import U256__hash__
+from ethereum.cancun.fork_types import (
+    MappingAddressAccount,
+    MappingAddressAccountStruct,
+    ListHash32__hash__,
+    ListHash32,
+    Address,
+    OptionalAddress,
+    SetAddress,
+    SetAddressDictAccess,
+    SetAddressStruct,
+    SetTupleAddressBytes32,
+    SetTupleAddressBytes32DictAccess,
+    SetTupleAddressBytes32Struct,
+    OptionalMappingAddressBytes32,
+    MappingAddressBytes32Struct,
+    TupleVersionedHash,
+    TupleVersionedHashStruct,
+    VersionedHash,
+)
+from ethereum.cancun.transactions_types import To, ToStruct
+
+from cairo_core.bytes_impl import Bytes32__hash__
+from cairo_core.hash.blake2s import blake2s_add_uint256, blake2s, blake2s_add_felt
+
+from mpt.trie_diff import OptionalUnionInternalNodeExtendedImpl
+
+from mpt.hash_diff import (
+    hash_state_storage_diff,
+    hash_state_account_diff,
+    hash_account_diff_segment,
+    hash_storage_diff_segment,
+)
+from mpt.types import (
+    NodeStore,
+    OptionalUnionInternalNodeExtended,
+    MappingBytes32Bytes32,
+    MappingBytes32Address,
+)
+from mpt.trie_diff import compute_diff_entrypoint
+from mpt.utils import sort_account_diff, sort_storage_diff
+
+func main{
+    output_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr,
+    ecdsa_ptr: SignatureBuiltin*,
+    bitwise_ptr: BitwiseBuiltin*,
+    ec_op_ptr: EcOpBuiltin*,
+    keccak_ptr: felt*,
+    poseidon_ptr: PoseidonBuiltin*,
+    range_check96_ptr: felt*,
+    add_mod_ptr: ModBuiltin*,
+    mul_mod_ptr: ModBuiltin*,
+}() {
+    alloc_locals;
+
+    // Program inputs from the init.cairo program.
+    local block: Block;
+    local chain: BlockChain;
+    local withdrawals_trie: TrieBytesOptionalUnionBytesWithdrawal;
+
+    // Program inputs from the body.cairo program.
+    local block_transactions: TupleUnionBytesLegacyTransaction;
+    local state: State;
+    local transactions_trie: TrieBytesOptionalUnionBytesLegacyTransaction;
+    local receipts_trie: TrieBytesOptionalUnionBytesReceipt;
+    local block_logs: TupleLog;
+    local block_hashes: ListHash32;
+    local gas_available: Uint;
+    local chain_id: U64;
+    local blob_gas_used: Uint;
+    local excess_blob_gas: U64;
+
+    // MPT diffs inputs
+    local node_store: NodeStore;
+    local address_preimages: MappingBytes32Address;
+    local storage_key_preimages: MappingBytes32Bytes32;
+    local post_state_root: OptionalUnionInternalNodeExtended;
+
+    // Fill-in the program inputs through the hints.
+    %{ teardown_inputs %}
+
+    // // Because in args_gen we want to generate a state with (prev, new) tuples, we pass an initial snapshot of the state.
```

Comment:
should we do this in args_gen instead of in the cairo code?

Reply:
we don't have a way to do this in args gen unfortunately